hyperon_das_atomdb/adapters/ram_only.py

```diff
@@ -71,53 +84,85 @@ def _add_atom_type(self, _name: str, _type: Optional[str] = 'Type'):
+        self.db.atom_type.pop(key, None)
+        self.all_named_types.remove(_name)
+
     def _add_outgoing_set(self, key: str, targets_hash: Dict[str, Any]) -> None:
         self.db.outgoing_set[key] = targets_hash
 
-    def _add_incomming_set(self, key: str, targets_hash: Dict[str, Any]) -> None:
+    def _get_and_delete_outgoing_set(self, handle: str) -> List[str]:
+        return self.db.outgoing_set.pop(handle, None)
+
+    def _add_incoming_set(self, key: str, targets_hash: Dict[str, Any]) -> None:
         for target_hash in targets_hash:
-            incomming_set = self.db.incomming_set.get(target_hash)
-            if incomming_set is None:
-                self.db.incomming_set[target_hash] = [key]
+            incoming_set = self.db.incoming_set.get(target_hash)
+            if incoming_set is None:
+                self.db.incoming_set[target_hash] = [key]
             else:
-                self.db.incomming_set[target_hash].append(key)
+                self.db.incoming_set[target_hash].append(key)
+
+    def _delete_incoming_set(self, link_handle: str, atoms_handle: List[str]) -> None:
+        for atom_handle in atoms_handle:
+            handles = self.db.incoming_set.get(atom_handle, [])
+            if len(handles) > 0:
+                handles.remove(link_handle)
+            if len(handles) == 0:
+                self.db.incoming_set.pop(atom_handle, None)
 
     def _add_templates(
-        self,
-        composite_type_hash: str,
-        named_type_hash: str,
-        key: str,
-        targets_hash: List[str],
+        self, composite_type_hash: str, named_type_hash: str, key: str, targets_hash: List[str]
     ) -> None:
         template_composite_type_hash = self.db.templates.get(composite_type_hash)
         template_named_type_hash = self.db.templates.get(named_type_hash)
 
         if template_composite_type_hash is not None:
-            # template_composite_type_hash.append([key, targets_hash])
             template_composite_type_hash.append((key, tuple(targets_hash)))
         else:
-            # self.db.templates[composite_type_hash] = [[key, targets_hash]]
             self.db.templates[composite_type_hash] = [(key, tuple(targets_hash))]
 
         if template_named_type_hash is not None:
-            # template_named_type_hash.append([key, targets_hash])
             template_named_type_hash.append((key, tuple(targets_hash)))
         else:
-            # self.db.templates[named_type_hash] = [[key, targets_hash]]
             self.db.templates[named_type_hash] = [(key, tuple(targets_hash))]
 
+    def _delete_templates(self, link_document: dict, targets_hash: List[str]) -> None:
+        template_composite_type = self.db.templates.get(link_document['composite_type_hash'], [])
+        if len(template_composite_type) > 0:
+            template_composite_type.remove((link_document['_id'], tuple(targets_hash)))
+        if len(template_composite_type) == 0:
```

Comment:
This is wrong. The template key is still searchable even if there's no atom that matches it. We should leave the empty list.

Reply:
done